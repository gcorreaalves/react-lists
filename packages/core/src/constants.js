export const PACKAGE_NAME = 'react-chat'
