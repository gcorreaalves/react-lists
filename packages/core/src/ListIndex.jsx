import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles'
import { PACKAGE_NAME } from './constants'

const ListIndex = ({ height, items, onItemClick }) => {
  const padding = (height / items.length - 14) / 2
  const itemStyles = {
    ...styles.indexMenuListItem,
    padding
  }

  const renderItem = key => (
    <li key={key} style={itemStyles}>
      <a
        onClick={onItemClick(`#${PACKAGE_NAME}-index-${key.toLowerCase()}`)}
        style={styles.indexMenuListItemLink}
      >
        {key}
      </a>
    </li>
  )

  if (items.length) {
    return (
      <nav style={styles.indexMenu}>
        <ul style={styles.indexMenuList}>{items.map(renderItem)}</ul>
      </nav>
    )
  }

  return null
}

ListIndex.propTypes = {
  height: PropTypes.number.isRequired,
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  onItemClick: PropTypes.func.isRequired
}

export default ListIndex
