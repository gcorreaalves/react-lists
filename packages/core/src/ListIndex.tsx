import React from 'react'
import { PACKAGE_NAME } from './constants'

const styles: { [key: string]: React.CSSProperties } = {
  root: {
    position: 'absolute',
    right: 20,
    top: 0,
    zIndex: 2
  },
  list: {
    listStyle: 'none',
    padding: 0,
    margin: 0,
    width: '100%'
  },
  listItem: {
    lineHeight: 1,
    textAlign: 'center'
  },
  listItemLink: {
    color: 'rgba(0, 0, 0, 0.54)',
    fontSize: '14px',
    cursor: 'pointer',
    textDecoration: 'none'
  }
}

interface IProps {
  height: number
  items: Array<string>
  onItemClick: (selector: string) => (event: React.MouseEvent) => void
}

const ListIndex: React.FunctionComponent<IProps> = ({
  height,
  items,
  onItemClick
}) => {
  const padding = (height / items.length - 14) / 2
  const itemStyles = {
    ...styles.listItem,
    padding
  }

  const renderItem = key => (
    <li key={key} style={itemStyles}>
      <a
        onClick={onItemClick(`#${PACKAGE_NAME}-index-${key.toLowerCase()}`)}
        style={styles.listItemLink}
      >
        {key}
      </a>
    </li>
  )

  if (items.length) {
    return (
      <nav style={styles.root}>
        <ul style={styles.list}>{items.map(renderItem)}</ul>
      </nav>
    )
  }

  return null
}

export default ListIndex
