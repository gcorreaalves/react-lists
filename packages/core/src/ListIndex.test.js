import React from 'react'
import { mount, render } from 'enzyme'
import ListIndex from './ListIndex'
import { PACKAGE_NAME } from './constants'

describe('ListIndex component', () => {
  it('should render correctly', () => {
    const wrapper = render(
      <ListIndex
        height={100}
        items={['A', 'B', 'C']}
        onItemClick={() => {
          return null
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render 3 items', () => {
    const wrapper = mount(
      <ListIndex
        height={100}
        items={['A', 'B', 'C']}
        onItemClick={() => {
          return null
        }}
      />
    )

    expect(wrapper.find('li').length).toBe(3)
  })

  it('should render empty', () => {
    const wrapper = mount(
      <ListIndex
        height={100}
        items={[]}
        onItemClick={() => {
          return null
        }}
      />
    )

    expect(wrapper.find('li').length).toBe(0)
  })

  it('should call onItemClick', () => {
    const handleClick = jest.fn()
    const wrapper = mount(
      <ListIndex
        height={100}
        items={['A', 'B', 'C']}
        onItemClick={handleClick}
      />
    )

    wrapper.find('li').first().find('a').simulate('click')

    expect(handleClick).toHaveBeenCalledWith(`#${PACKAGE_NAME}-index-a`)
  })
})
