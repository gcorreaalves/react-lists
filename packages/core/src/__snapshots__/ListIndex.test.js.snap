// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListIndex component should render correctly 1`] = `
<nav
  style="position:absolute;right:20px;top:0;z-index:2"
>
  <ul
    style="list-style:none;padding:0;margin:0;width:100%"
  >
    <li
      style="line-height:1;text-align:center;padding:9.666666666666668px"
    >
      <a
        style="color:rgba(0, 0, 0, 0.54);font-size:14px;cursor:pointer;text-decoration:none"
      >
        A
      </a>
    </li>
    <li
      style="line-height:1;text-align:center;padding:9.666666666666668px"
    >
      <a
        style="color:rgba(0, 0, 0, 0.54);font-size:14px;cursor:pointer;text-decoration:none"
      >
        B
      </a>
    </li>
    <li
      style="line-height:1;text-align:center;padding:9.666666666666668px"
    >
      <a
        style="color:rgba(0, 0, 0, 0.54);font-size:14px;cursor:pointer;text-decoration:none"
      >
        C
      </a>
    </li>
  </ul>
</nav>
`;
